more compiler cleanup [nfc]

vtjnash opened the pull request.

vtjnash committed:
codegen: remove dead code for arrayvar optimization

vtjnash committed:
compiler warnings: squelch them

vtjnash commented:
@nanosoldier `runbenchmarks(ALL, vs = ":master")`

StefanKarpinski commented:
What does "nfc" stand for? Near-field communication?

Keno commented:
No functional changes

StefanKarpinski commented on src/jloptions.c:
```diff
@@ -31,12 +31,12 @@ jl_options_t jl_options = { 0,    // quiet
                             NULL, // cpu_target ("native", "core2", etc...)
                             0,    // nprocs
                             NULL, // machine_file
+                            NULL, // project
```
Thanks for the fix!

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/27945#issuecomment-403058732) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/afe8465b91c6a995b18577a2eaabec737e47e851/2ba2025_vs_feaee9e/report.md). cc @ararslan

vtjnash commented:
I don't understand how this could have affected allocations for `sumcolon_view` and `sumrange_view `, and time for `["problem", "seismic"]`

vtjnash commented:
@nanosoldier `runbenchmarks("sumcolon_view" || "sumrange_view" || "seismic", vs = ":master")`

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/27945#issuecomment-403255562) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/aa0b1dcbd345d743f3958fbfcf83b1de514f26c3/330d857_vs_2245a04/report.md). cc @ararslan

vtjnash committed:
codegen: fix optimization misses from IR changes

- dereferenceable attribute requires align attribute (if eltype is not
sized) or it is simply ignored

- tbaa_const isn't necessarily sufficient to allow LICM hoisting. an
unknown call instruction (such as julia.gc_alloc_obj) present in the
same loop can prevent TBAA from declaring the memory to be immutable.
However, the invariant.load attribute is much stronger:
additionally mark all tbaa_const loads with MD_immutable_load,
permitting llvm to reorder and left them them much more freely
(requiring only dereferenceable).

vtjnash commented:
Not precisely NFC anymore, since this uncovered optimizations that we thought were happening but weren’t, so @nanosoldier `runbenchmarks(ALL, vs = ":master")` with the fixes for those

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/27945#issuecomment-403677917) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/4bd63b6ad761299bce29f706a1eff2c1e4cec3df/1ec85fb_vs_d8f70a5/report.md). cc @ararslan
